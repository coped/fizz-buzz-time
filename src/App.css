.App {
  text-align: center;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.bold {
  font-weight: bold;
}

/* this class is for testing purposes */
.test {
  color: blue;
}