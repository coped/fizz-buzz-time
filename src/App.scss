#App {
  display: flex;
  height: 100vh;
  width: 100vw;

  .bold {
    font-weight: bold;
  }

  .button {
    border: 2px solid #707070;
    color: #707070;
    padding: 15px 30px;
    text-align: center;
    font-size: 16px;
    font-weight: bold;
    cursor: pointer;
    border-radius: 5px;
  }

  .description {
    font-size: 30px;
  }

  /* this class is for testing purposes */
  .test {
    color: blue;
  }

  /* Transition classes for ReactCSSTransitionGroup */
  .example-enter {
    opacity: 0.01;
  }

  .example-enter.example-enter-active {
    opacity: 1;
    transition: opacity 500ms ease-in;
  }

  .example-leave {
    opacity: 1;
  }

  .example-leave.example-leave-active {
    opacity: 0.01;
    transition: opacity 300ms ease-in;
  }
}
