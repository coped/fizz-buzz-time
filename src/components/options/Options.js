import React from "react";
import "./Options.scss";

export default function Options(props) {
  const { toggleShowTimer, handleChange, fizz, buzz, count } = props;
  const isDisabled = count > 0 ? true : false;
  return (
    <div id="Options">
      <p className="description">
        Please enter a fizz and buzz time in seconds.{" "}
        <strong>Values should be 2 to 10, inclusive.</strong>
      </p>
      <div className="fizz-buzz-fields">
        <label id="fizz-label" htmlFor="fizz" className="label">
          Fizz:
        </label>
        <input
          id="fizz"
          className="input"
          name="fizz"
          type="text"
          value={fizz}
          onChange={handleChange}
          disabled={isDisabled}
        />
        <label id="buzz-label" htmlFor="buzz" className="label">
          Buzz:
        </label>
        <input
          id="buzz"
          className="input"
          name="buzz"
          type="text"
          value={buzz}
          onChange={handleChange}
          disabled={isDisabled}
        />
      </div>
      <button
        id="to-timer"
        onClick={toggleShowTimer}
        type="button"
        className="button"
      >
        Go to timer >
      </button>
    </div>
  );
}
