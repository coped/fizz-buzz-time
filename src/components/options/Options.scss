#Options {
  display: flex;
  height: 100%;
  width: 100%;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  .button {
    margin-top: 50px;
  }

  .fizz-buzz-fields {
    margin-top: 40px;
  }

  .label {
    font-size: 18px;
    margin-right: 5px;
  }

  .input {
    width: 100px;
    padding: 8px 16px;
    border: 1px solid #707070;
    background-color: #f5f5f5;
    border-radius: 5px;
    font-size: 16px;
    margin-right: 40px;
  }
}
