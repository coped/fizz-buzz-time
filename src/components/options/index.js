import Options from "components/options/Options";

export { Options };
