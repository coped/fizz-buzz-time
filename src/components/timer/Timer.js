import React, { Component } from "react";
import { Button } from "components/common";

export default class Timer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      something: "something",
    };
  }
  render() {
    const { toggleShowTimer } = this.props;
    return (
      <div id="Timer">
        <Button onClick={toggleShowTimer}>{"< Set times"}</Button>
        <p>I am a timer!</p>
      </div>
    );
  }
}
