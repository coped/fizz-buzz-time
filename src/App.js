import React, { Component } from "react";
import "App.scss";
import { Timer } from "components/timer";
import { Options } from "components/options";

export default class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showTimer: false,
      count: 0,
      counting: false,
      fizz: "",
      buzz: "",
    };

    this.toggleShowTimer = this.toggleShowTimer.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.startTimer = this.startTimer.bind(this);
    this.stopTimer = this.stopTimer.bind(this);
    this.incrementCount = this.incrementCount.bind(this);
  }

  // Lifecycle methods

  componentWillUnmount() {
    clearInterval(this.countInterval);
  }

  // Component methods

  startTimer() {
    if (!this.state.counting) {
      this.countInterval = setInterval(this.incrementCount, 1000);
      this.setState({ counting: true });
    }
  }

  stopTimer() {
    if (this.state.counting) {
      clearInterval(this.countInterval);
      this.setState({ counting: false });
    } else {
      this.setState({ count: 0 });
    }
  }

  incrementCount() {
    if (this.state.count >= 35999) this.setState({ count: 0 });
    this.setState({ count: this.state.count + 1 });
  }

  toggleShowTimer() {
    this.setState({ showTimer: !this.state.showTimer });
  }

  handleChange(event) {
    const target = event.target;

    this.setState({
      [target.name]: target.value.replace(/\D/, ""),
    });
  }

  showTimerOrOptions(showTimer) {
    const { count, counting, fizz, buzz } = this.state;
    if (showTimer) {
      return (
        <Timer
          key="timer"
          toggleShowTimer={this.toggleShowTimer}
          count={count}
          counting={counting}
          startTimer={this.startTimer}
          stopTimer={this.stopTimer}
          fizz={fizz}
          buzz={buzz}
        />
      );
    } else {
      return (
        <Options
          key="options"
          toggleShowTimer={this.toggleShowTimer}
          handleChange={this.handleChange}
          count={count}
          fizz={fizz}
          buzz={buzz}
        />
      );
    }
  }
  render() {
    return <div id="App">{this.showTimerOrOptions(this.state.showTimer)}</div>;
  }
}
