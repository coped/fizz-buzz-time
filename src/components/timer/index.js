import Timer from "components/timer/Timer"

export { Timer }