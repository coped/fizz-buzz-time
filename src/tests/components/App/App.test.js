import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import App from "App";
import { Options } from "components/options";
import { Timer } from "components/timer";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

Enzyme.configure({ adapter: new Adapter() });

describe("App", () => {
  const mockTimes = {
    starting: "0:00:00",
    seconds: "0:00:11",
    minutes: "0:20:17",
    hours: "2:20:19",
  };
  const selectors = {
    startTimer: "#start-timer",
    stopTimer: "#stop-timer",
    timerOutput: "#timer-output",
    fizzBuzzOutput: "#fizz-buzz-output",
    fizz: "#fizz",
    buzz: "#buzz",
  };

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders options", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(Options).length).toEqual(1);
  });

  it("toggles between options and timer", () => {
    const wrapper = mount(<App />);
    wrapper.find("#to-timer").simulate("click");
    expect(wrapper.find(Timer).length).toEqual(1);
    wrapper.find("#to-options").simulate("click");
    expect(wrapper.find(Options).length).toEqual(1);
  });

  describe("startTimer()", () => {
    it("starts the timer", () => {
      jest.useFakeTimers();
      const wrapper = mount(<App />);
      wrapper.setState({ showTimer: true });
      const button = wrapper.find(selectors.startTimer);
      button.simulate("click");
      expect(wrapper.state().counting).toEqual(true);
      expect(wrapper.state().count).toEqual(0);
      // Timer count should increase by 1 per second
      jest.advanceTimersByTime(1000);
      expect(wrapper.state().count).toEqual(1);
      jest.advanceTimersByTime(1000);
      expect(wrapper.state().count).toEqual(2);
    });
  });

  describe("stopTimer()", () => {
    it("stops the timer", () => {
      jest.useFakeTimers();
      const wrapper = mount(<App />);
      wrapper.setState({ showTimer: true });
      const startButton = wrapper.find(selectors.startTimer);
      const stopButton = wrapper.find(selectors.stopTimer);
      startButton.simulate("click");
      // Timer count advances to 1
      jest.advanceTimersByTime(1000);
      // Stop timer
      stopButton.simulate("click");
      expect(wrapper.state().counting).toEqual(false);
      // After 5 seconds, count should still be at 1
      jest.advanceTimersByTime(5000);
      expect(wrapper.state().count).toEqual(1);
    });

    it("resets the timer once stopped", () => {
      jest.useFakeTimers();
      const wrapper = mount(<App />);
      wrapper.setState({ showTimer: true });
      const startButton = wrapper.find(selectors.startTimer);
      const stopButton = wrapper.find(selectors.stopTimer);
      startButton.simulate("click");
      // Timer count advances to 1;
      jest.advanceTimersByTime(1000);
      // Stop timer
      stopButton.simulate("click");
      // Pressing stop button while not counting should reset count
      stopButton.simulate("click");
      expect(wrapper.state().count).toEqual(0);
    });
  });

  describe("timer output", () => {
    it("outputs the progression of elapsed time", () => {
      jest.useFakeTimers();
      const wrapper = mount(<App />);
      wrapper.setState({ showTimer: true });
      const output = wrapper.find(selectors.timerOutput);
      expect(output.text()).toEqual(mockTimes.starting);
      wrapper.find(selectors.startTimer).simulate("click");
      jest.advanceTimersByTime(11000);
      expect(output.text()).toEqual(mockTimes.seconds);
    });

    it("should only show output at appropriate times", () => {
      jest.useFakeTimers();
      const wrapper = mount(<App />);
      wrapper.setState({ showTimer: true });
      const output = wrapper.find(selectors.fizzBuzzOutput);
      wrapper.setState({ fizz: 3, buzz: 5, count: 14 });
      expect(output.text()).toEqual("");
      // Start advancing time
      wrapper.find(selectors.startTimer).simulate("click");
      jest.advanceTimersByTime(1000);
      expect(output.text()).toEqual("FizzBuzz");
      jest.advanceTimersByTime(1000);
      expect(output.text()).toEqual("");
    });
  });

  describe("options", () => {
    it("modifies fizz values", () => {
      const wrapper = mount(<App />);
      wrapper.find(selectors.fizz).simulate("change", {
        target: {
          name: "fizz",
          value: "3",
        },
      });
      expect(wrapper.state().fizz).toEqual("3");
    });

    it("modifies buzz values", () => {
      const wrapper = mount(<App />);
      wrapper.find(selectors.buzz).simulate("change", {
        target: {
          name: "buzz",
          value: "5",
        },
      });
      expect(wrapper.state().buzz).toEqual("5");
    });

    it("disables fizz input once time has progressed", () => {
      const wrapper = mount(<App />);
      const fizzInput = wrapper.find(selectors.fizz);
      expect(fizzInput.getDOMNode().disabled).toEqual(false);
      wrapper.setState({ count: 1 });
      expect(fizzInput.getDOMNode().disabled).toEqual(true);
    });

    it("disables buzz input once time has progressed", () => {
      const wrapper = mount(<App />);
      const buzzInput = wrapper.find(selectors.buzz);
      expect(buzzInput.getDOMNode().disabled).toEqual(false);
      wrapper.setState({ count: 1 });
      expect(buzzInput.getDOMNode().disabled).toEqual(true);
    });
  });

  it("has a valid snapshot", () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
