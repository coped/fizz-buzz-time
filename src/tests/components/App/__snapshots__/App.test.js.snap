// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  id="App"
>
  <div
    id="Options"
  >
    <p
      className="description"
    >
      Please enter a fizz and buzz time in seconds.
       
      <strong>
        Values should be 2 to 10, inclusive.
      </strong>
    </p>
    <div
      className="fizz-buzz-fields"
    >
      <label
        className="label"
        htmlFor="fizz"
        id="fizz-label"
      >
        Fizz:
      </label>
      <input
        className="input"
        disabled={false}
        id="fizz"
        name="fizz"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        className="label"
        htmlFor="buzz"
        id="buzz-label"
      >
        Buzz:
      </label>
      <input
        className="input"
        disabled={false}
        id="buzz"
        name="buzz"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <button
      className="button"
      id="to-timer"
      onClick={[Function]}
      type="button"
    >
      Go to timer &gt;
    </button>
  </div>
</div>
`;
